src/ticks/exp_engine.c

```diff
@@ -0,0 +1,517 @@
+            char tp[128];
+            expression_result_type_to_string(&a->type, is_expression_signed(a), tp);
+            sprintf(result->as_error, "Cannot perform math '+' on type %s", tp);
+            break;
+        }
+    }
+}
+
+void expression_math_mul(struct expression_result_t* a, struct expression_result_t* b, struct expression_result_t* result) {
+    if (!(are_exp_same_type(a, &a->type, b, &b->type))) {
+        struct expression_result_t local_3;
+        convert_expression(b, &local_3, a->type.type_, is_expression_signed(b));
+        expression_math_mul(a, &local_3, result);
+        return;
+    }
+
+    switch (a->type.type_) {
+        case TYPE_FLOAT: {
+            result->as_float = a->as_float * b->as_float;
+            return;
+        }
+        case TYPE_CHAR:
+        case TYPE_INT:
+        case TYPE_LONG: {
+            if (is_expression_signed(a)) {
+                result->as_int = a->as_int * b->as_int;
+            } else {
+                result->as_uint = (uint32_t)((uint32_t)a->as_int * (uint32_t)b->as_int);
+            }
+            return;
+        }
+        default: {
+            result->flags |= EXPRESSION_ERROR;
+            char tp[128];
+            expression_result_type_to_string(&a->type, is_expression_signed(a), tp);
+            sprintf(result->as_error, "Cannot perform math '+' on type %s", tp);
+            break;
+        }
+    }
+}
+
+void expression_math_div(struct expression_result_t* a, struct expression_result_t* b, struct expression_result_t* result) {
+    if (!(are_exp_same_type(a, &a->type, b, &b->type))) {
+        struct expression_result_t local_3;
+        convert_expression(b, &local_3, a->type.type_, is_expression_signed(b));
+        expression_math_div(a, &local_3, result);
+        return;
+    }
+
+    switch (a->type.type_) {
+        case TYPE_FLOAT: {
+            result->as_float = a->as_float / b->as_float;
+            return;
+        }
+        case TYPE_CHAR:
+        case TYPE_INT:
+        case TYPE_LONG: {
+            if (is_expression_signed(a)) {
+                result->as_int = a->as_int / b->as_int;
+            } else {
+                result->as_uint = (uint32_t)((uint32_t)a->as_int / (uint32_t)b->as_int);
+            }
+            return;
+        }
+        default: {
+            result->flags |= EXPRESSION_ERROR;
+            char tp[128];
+            expression_result_type_to_string(&a->type, is_expression_signed(a), tp);
+            sprintf(result->as_error, "Cannot perform math '+' on type %s", tp);
+            break;
+        }
+    }
+}
+
+uint8_t is_primitive_integer_type(struct expression_result_t* exp) {
+    switch (exp->type.type_) {
+        case TYPE_CHAR:
+        case TYPE_INT:
+        case TYPE_LONG: {
+            return 1;
+        }
+        default: {
+            return 0;
+        }
+    }
+}
+
+void expression_result_value_to_string(struct expression_result_t* result, char* buffer) {
```

Comment:
debug.c::debug_print_element has got very similar printing logic - can we have just one copy of the code?

Reply:
Sadly, no, as expression_result_t is a separate entity. I can convert whatever in a `debug_print_element` by calling `debug_resolve_expression_element` into an `expression_result_t` first and then call `expression_result_value_to_string`, what do you think?

Reply:
Ugh, `debug_print_element` has this offset thing going on that really complicates things...

Reply:
Ah yeah. Makes me think we should use UT_string? It’s just trying to build up a buffer and not overflow ot.

Reply:
Let me combine the two somehow

Reply:
Ok, figured how make it to work. Wasn't able to test array thoroughly, because CDB generates a line which parser doesn't like (character `x` in particular)
```
Warning cannot parse type record.
Warning: could not add debug info on symbol.
Decoded cdb: <S:Fmain.c$haha55$0_0$0({4}DA2x,SI:S),E,0,0>
```
I think I should rewrite the parser anyway using yacc/bison, and fix arrayys there